Save and Restore config back-up files #42

phillxnet opened the pull request:
1. Copy static/config-backups/* to config-backups-rpmsave: incoming package %pre scriptlet (first to run).
2. Copy config-backups-rpmsave/* to static/config-backups: incoming package %posttrans (last to run and after build.sh).

It is important that we restore our config backup files after build.sh as it depends on a non-existent static directory to establish incoming static contents: see %post, in update mode, which ensures the removal of the static directory.

Fixes #42 

Forum reference for the following rockstor-core superset issue: https://github.com/rockstor/rockstor-core/issues/2639
https://forum.rockstor.com/t/unknown-internal-error-doing-a-get-to-api-config-backup/8938
Please update that thread with this partial address of the issues raised there.

phillxnet commented:
## Testing

### Install with no /opt/rockstor directory.
```
tail -n1 /var/log/zypp/history
# No config-backups-rpmsave directory found.
```
### Updating from 5.0.2-0 rpm with no config saves
```
tail -n1 /var/log/zypp/history
# No config-backups-rpmsave directory found.
```
### Updating from 5.0.2-0 rpm with 2 config saves in place
- Before
```
ls -la /opt/rockstor/static/config-backups/
total 8
drwxr-xr-x 1 root root 128 Aug 18 18:41 .
drwxr-xr-x 1 root root  94 Aug 18 18:40 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz
ls -la /opt/rockstor/config-backups-rpmsave
ls: cannot access '/opt/rockstor/config-backups-rpmsave': No such file or directory
```
- After 
```
tail -n1 /var/log/zypp/history
# Restoring config-backups-rpmsave contents to static/config-backups.

rleap15-5:/opt # ls -la /opt/rockstor/static/config-backups/
total 8
drwxr-xr-x 1 root root 128 Aug 18 18:46 .
drwxr-xr-x 1 root root  94 Aug 18 18:46 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz

ls -la /opt/rockstor/config-backups-rpmsave
total 12
drwxr-xr-x 1 root root 128 Aug 18 18:45 .
drwxr-xr-x 1 root root 456 Aug 18 18:46 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz

Web-UI System -> Config backups Displays as expected.

Add another config save via "Backup current config" button.
ls -la /opt/rockstor/static/config-backups/
total 12
drwxr-xr-x 1 root root 192 Aug 18 18:53 .
drwxr-xr-x 1 root root  94 Aug 18 18:46 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz
-rw-r--r-- 1 root root 755 Aug 18 18:53 backup-2023-08-18-185306.json.gz

ls -la /opt/rockstor/config-backups-rpmsave
total 12
drwxr-xr-x 1 root root 128 Aug 18 18:45 .
drwxr-xr-x 1 root root 560 Aug 18 18:51 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz

Reinstall the same rpm via cli:
zypper in -f rockstor
tail -n1 /var/log/zypp/history
# Restoring config-backups-rpmsave contents to static/config-backups.
- Restart our top level service as we are doing cli reinstall (or just reboot):
systemctl start rockstor-bootstrap

ls -la /opt/rockstor/static/config-backups/
total 12
drwxr-xr-x 1 root root 192 Aug 18 18:57 .
drwxr-xr-x 1 root root  94 Aug 18 18:57 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz
-rw-r--r-- 1 root root 755 Aug 18 18:53 backup-2023-08-18-185306.json.gz

ls -la /opt/rockstor/config-backups-rpmsave
total 16
drwxr-xr-x 1 root root 192 Aug 18 18:56 .
drwxr-xr-x 1 root root 560 Aug 18 18:58 ..
-rw-r--r-- 1 root root 753 Aug 18 18:40 backup-2023-08-18-184056.json.gz
-rw-r--r-- 1 root root 751 Aug 18 18:41 backup-2023-08-18-184103.json.gz
-rw-r--r-- 1 root root 755 Aug 18 18:53 backup-2023-08-18-185306.json.gz

Web-UI System -> Config backups Displays as expected.
```

Hooverdan96 commented:
@phillxnet, nice going!
Question though, you're copying the files back to the original location (if they existed before the update). Should you move them instead (or copy and then delete the contents of `config-backups-rpmsave`)? So, if the next update comes you don't run into a conflict on overwriting the files in the `config-backups-rpmsave` folder or copying back "old" configs that have been deleted by the user since the last update? Or does the rpmbuild clean it up automatically already somewhere else?

phillxnet commented:
> So, if the next update comes you don't run into a conflict on overwriting the files in the config-backups-rpmsave folder or copying back "old" configs that have been deleted by the user since the last update?

I tested for this failure in the reported test where I added a third config backup. Given the following:

Here I am banking - assuming that on each back-up file remains unchanged. This is a safe assumption as we only offer, within Web-UI the chance to create new backups with unique (date-time stamped) names. So there is no 'in-place' modifications of any existing files. The db also assumes this givne it's md5 on these files: hence the archive nature of the copy used (preserve=all).

We do definitely here have an unending build-up of all prior config-backup files: but at this level we have no means of knowing which is within the db. We are just a simple copy past to preserve what has been. If prior files have been wiped in db (Web-UI) their return is not a problem as I understand it. Not clean but still an improvement on what we have currently: complete loss of all files. I had considered an rsync arrangement: and that could be an upgrade in this situation. But that is not preserving files that folks may have accidentally deleted: where-as this brute-force approach does.

So we are first-pass on this one hence the simplest solution first. We can fines later once we have some field testing of what we have here was my thinking. Plus the more we second guess what folks want to save he more we may put ourselves into a corner.

> Or does the rpmbuild clean it up automatically already somewhere else?

No, currently this exist outside the rpm's clean-up set. It never actually owns these directories or files: given they are configuration dependent. Kind of like the .initrock: we never touch it - and file to wipe the install directory if there is anything left behind that was install confing generated like the certs. The exception is the pyc files which we wipe anyway, but they are auto regenerated in all cases anyway.

So yes - build-up - but this is a back-up/archive type arrangement where all prior config-backup files are 'carried' and maintained throughout the lifetime of the install. Irrispective of a Web-UI delete.

A follow-up feature to this 'invisible history' would be a Web-UI import mechanism: where we scan the currently config-back store for prior (preserved via this back-up method) or just by-hand placed configurations: outside the scope of this issue/pr set however. That same rescan mechanism would also pick-up by-hand modifications of existing config-backup files that I think would currently be rejected due to consequent md5 miss-match imparted to altered files. Just a thought.

Hooverdan96 commented:
@phillxnet thanks for the detailed explanation of your thought process. Certainly agreed that this is an improvement and we can refine later.

I do disagree with the notion of eternal build-up of the config files though. My argument, based on the below scenario, is this:
- fresh install of latest version of Rockstor , e.g. 4.1.0
- say, I have accumulated 5 configuration backups before the next upgrade to 4.2.0 comes along
- during the update, all 5 configs are copied to `config-backups-rpmsave`, the `config-backups` is wiped, recreated and the 5 backups are placed back in `config-backups-rpmsave`. The 5 files in `config-backups-rpmsave` remain
- I accumulate a few more backups, say, 3 to a total of 8. I delete the three oldest using the WebUI, i.e. database and directory are cleansed of those.
- I now update to the next version 4.3.0. 
- during the update, the 5 backups (3 since 4.2 and 2 since 4.1) are copied to `config-backups-rpmsave`, the `config-backups` is wiped, recreated and the 5 files just moved are copied back in `config-backups-rpmsave` (like you said, they would likely not have been meddled with anyway, so overwriting anything is not a problem) but in addition the 3 that I had explicitly deleted between 4. 2 and 4.3 are copied over as well. So, while the database now has a record of the 5 files and their corresponding physical files, the 3 files that I presumed I had deleted at some earlier time are sitting in the shadows there.
- My user expectation is that my "state" should be the same as before I upgraded from 4.2 to 4.3, i.e. only the 5 files (3 new, 2 old) would end up in that directory (and database but that's already guaranteed throughout the update).

I don't think that would be an ambiguous expectation, since from the user perspective, in my opinion, we should expect that the status is preserved (i.e. contents of a directory with user generated files match before/after update, irrespective of that's done technically). So, I think that the contents of `config-backups-rpmsave` should be deleted after a successful copy back and the update is complete. That way we have a consistent state of the `config-backups` directory unaffected by an update.

With that said, it's still better than what we had before, which was nothing :).
Thanks again.

phillxnet commented:
@Hooverdan96 Thanks for the further thoughts on this one, in many ways I agree: and as you seem to appreciate we were after a fix here. I just thought we also had the opportunity to play ultra safe and delete no config-backups (kind of the opposite of what we were doing since moving Poetry): while potentially providing a history (preserving Web-UI removed) config backup files. Note that some folks will not even appreciate that these are files: and until they download them - they should also not consider them as saved (same machine and all that). The the more care - redundancy we can offer the better.

So all in this is all up for further development: as always. But to address your points in more detail we have the following.

> So, while the database now has a record of the 5 files and their corresponding physical files, the 3 files that I presumed I had deleted at some earlier time are sitting in the shadows there.

Yes, with the current fix, we do have this artefact. However there are a lot of files that serve functions not visible to the user - this is not necessarily a problem. And with the mentioned functional extension:

> So yes - build-up - but this is a back-up/archive type arrangement where all prior config-backup files are 'carried' and maintained throughout the lifetime of the install. Irrispective of a Web-UI delete.

&

> A follow-up feature to this 'invisible history' would be a Web-UI import mechanism: where we scan the current config-back store for prior (preserved via this back-up method) or just by-hand placed configurations: outside the scope of this issue/pr set however.

these could be resurfaced: with the cost of some trivial storage use.

> * My user expectation is that my "state" should be the same as before I upgraded from 4.2 to 4.3.

I agree, and what we end up with does exactly this. We do not meddle with the db with this fix: just restore what the db expected to be there (along with some accumulated history). The user experience also does not see that our database also has an ongoing - also never removed - history of drive usage over the lifetime of an install: but it does (at least from memory I think it does :) ). And when this is eventually surfaced within the Web-UI, I expect folks will be pleased, and appreciate the history going back before the Web-UI surfacing of the same data that was not appreciated earlier (or know about): as it was not surfaced. We may have the same opportunity here. The proposed follow-up (turn around) of an import function will be a necessity when/if we come to a configurable location feature for these files. Now outstanding for over a decade. I.e. "I store my config-backup files here, but you are a new install so I appreciate/need your facility to import my old config-backup files - on request: I see your warning that these files cannot be verified as they were not created by the current install." type thing.

Again: user observed state is far smaller than system state in any make-easy approach: such as our appliance objective. And we can produce magic/welcome results if we play the line between exposure and hidden features properly.

> since from the user perspective, in my opinion, we should expect that the status is preserved

Agreed; and it is - functionally. The Web-UI looks the same, there is no functional difference other than our potential ability to re-surface backed-up config from the past - with the above suggested warning during import. Again at negligible system resource cost. With the win (play-safe) that we never delete anything in this very low level event during the rpm update.

> (i.e. contents of a directory with user generated files match before/after update, irrespective of that's done technically)

But these are not user generated - they result from user requests: like db updates do (e.g. the currently hidden storage history within the db). That is the difference in perception here - these files are the remit of our system and we previously lost them - we now do not (but we both appreciate the fix element here obviously). We only disagree (a small amount) on the utility. The expectation and observation from user perspective is as  was: at least from before we moved to Poetry and did the correct think there-after of wipe and re-pave the static. Incidentally that is exactly what a collectstatic clean will also do: see an old bug report where this was not appreciated or documented:
https://code.djangoproject.com/ticket/22891
I was aware of this but simply forgot entirely that our config-backup files lived here - but did appreciate that moving them anywhere else would involve quite a few system enhancements on our part - especially if we throw in a user configurable directory such as was suggested as a work around for my technical faux pas.

> So, I think that the contents of config-backups-rpmsave should be deleted after a successful copy back and the update is complete. That way we have a consistent state of the config-backups directory unaffected by an update.

Clean solution to our potential conundrum here. How would be know if we have a successful install?

We are basically aligned here: and this patch presented what I felt was a minimal fix - with a harmless, but potentially usefull feature of retrieving accidentally Web-UI removed config-backups in it's current state.

But all-in-all I do not have any hard conviction here. If @FroggyFlox was of the opinion that we do the delete you suggest then that could be merged in our current testing cycle. But we could equally move to an rsync (with delete say) arrangement rather than the blind cp with all preserved that I submitted here: as a first pass, and urgent, fix.

> I had considered an rsync arrangement: and that could be an upgrade in this situation. But that is not preserving files that folks may have accidentally deleted: where-as this brute-force approach does.

So all in I'm obviously game for enhancements and fully acknowledge this is not the optimal solution for all things :). And very definitely represents an almost minimal approach to a fix that was required at the time. However I do stand by the potential enhancements we may loose: but agree entirely there is something uncouth about this first pass: and that is exactly what you say: we are not cleaning up after ourselves. But lets also not throw the baby out with the bath water here - some further consideration of what this may facilitate may be good. But at the same time folks likely just want their last or one-before-last config anyway. And import capabilities and retrieving past (thought removed/deleted/wiped whatever)  backup-configs - or frankly configurable save locations are all potentially small features that, as mentioned, have yet to surface in programmer effort to date. And have also not actually been requested to any significant level.

So I'm easy both ways really, if there is due consideration first. Here I simply presented the minimum solution to our deleting all of these files: that of deleting none of them - even the out-of-db ones. But the uncouth element here remains. I just wanted to re-iterate my take on not deleting and it's potential future facility. But again, I'm open to improvements if we settle that way in this discussion.

Quite the discussion for what is one of the smallest easiest fixes to date within this testing channel phase. All good and most welcome as always.

Again thanks for your thought and attention on these things. Always much appreciated. I did not think we had the time here to discuss this all first as it goes. And given we are early testing channel - we can have this all fixed one way or another before our next stable spin-off. That is the basic aim here. Can a stable release update to the next stable released - that is our ideal goal here. So we have time to re-address this: and if @FroggyFlox if happy we can very well drop in a wipe of our config-backups-rpm content - or directory - if that is the consensus. And it all works out as intended :) .

Hooverdan96 commented:
And you are acorrect, it's a long discussion for one of the smaller pieces of Rockstor ... not unimportant obviously, but with the current fix we at least have (like you said) a path forward.
